File: Part1DataStructureImplementation/Trie.java
package Part1DataStructureImplementation;

import java.util.*;
//time spend about 30min
public class Trie {
    TrieNode root;
    public Trie(){
        root = new TrieNode();
    }
    private void insert(String word){
        TrieNode curr = root;
        for(char c: word.toCharArray()){
            int index = c- 'a';
            if(curr.children[index] == null){
                curr.children[index] = new TrieNode();
            }
            curr = curr.children[index];
        }
        curr.validWord = true;
    }
    private boolean isValidWord(String word){
        TrieNode curr = root;
        for(char c : word.toCharArray()){
            int index = c- 'a';
            if(curr.children[index]==null){
                return false;
            }
            curr = curr.children[index];
        }
        return curr.validWord;
    }
    private void remove(String word){
        if(!isValidWord(word)){
    return;
}
        TrieNode curr = root;
        List<TrieNode> path = new ArrayList<>();
        path.add(curr);
        for(char c: word.toCharArray()){
            int index = c-'a';
          
                 path.add(curr.children[index]);
                curr = curr.children[index];
            
           
         
            
        }

        curr.validWord = false;
        
        for(int i=path.size()-1;i>0;i--){
            TrieNode current = path.get(i);
            TrieNode parent = path.get(i-1);
            if(!current.validWord && noChildren(current)){
                int index = word.charAt(i-1) - 'a';
               parent.children[index] = null;
            }else{
                break;
            }
        }
    }
    private boolean noChildren(TrieNode current){
            for(TrieNode curr : current.children){
                if(curr!=null){
                    return false;
                }
            }
            return true;
        }
        public static void main(String[] args){
           
    Trie trie = new Trie();

    trie.insert("cat");
    trie.insert("car");
    trie.insert("care");
    trie.insert("dog");

    System.out.println(trie.isValidWord("cat"));   // true
    System.out.println(trie.isValidWord("car"));   // true
    System.out.println(trie.isValidWord("care"));  // true
    System.out.println(trie.isValidWord("ca"));    // false
    System.out.println(trie.isValidWord("dog"));   // true
    System.out.println(trie.isValidWord("do"));    // false
    System.out.println(trie.isValidWord("cow"));   // false

    trie.remove("care");

    System.out.println("After removing care:");
    System.out.println(trie.isValidWord("care"));  // false
    System.out.println(trie.isValidWord("car"));   // true
    System.out.println(trie.isValidWord("cat"));   // true

    trie.remove("car");

    System.out.println("After removing car:");
    System.out.println(trie.isValidWord("car"));   // false
    System.out.println(trie.isValidWord("cat"));   // true

    trie.remove("cat");

    System.out.println("After removing cat:");
    System.out.println(trie.isValidWord("cat"));   // false
    System.out.println(trie.isValidWord("dog"));   // true

    trie.remove("notthere");

    System.out.println("After removing word not in trie:");
    System.out.println(trie.isValidWord("dog"));   // true
}
        }


 class TrieNode{
    TrieNode[] children;
    boolean validWord;
    public TrieNode(){
        children = new TrieNode[26];
        validWord = false;
    }
}

File: homework4/trie.py
# Part 1: Data Structure Implementation - Trie
# Time: O(L) per operation (L = word length)
# Space: O(total characters inserted)


class TrieNode:
    def __init__(self):
        self.children = {}
        self.validWord = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        curr = self.root
        for char in word:
            if char not in curr.children:
                curr.children[char] = TrieNode()
            curr = curr.children[char]
        curr.validWord = True

    def isValidWord(self, word):
        curr = self.root
        for char in word:
            if char not in curr.children:
                return False
            curr = curr.children[char]
        return curr.validWord

    def remove(self, word):
        self._removeHelper(self.root, word, 0)

    def _removeHelper(self, node, word, index):
        # returns True if this node is safe to delete
        if index == len(word):
            node.validWord = False
            return len(node.children) == 0

        char = word[index]
        if char not in node.children:
            return False

        if self._removeHelper(node.children[char], word, index + 1):
            del node.children[char]
            return len(node.children) == 0 and not node.validWord
        return False


trie = Trie()
trie.insert("cat")
trie.insert("car")
trie.insert("card")
print(trie.isValidWord("car"))   # True
print(trie.isValidWord("ca"))    # False
trie.remove("car")
print(trie.isValidWord("car"))   # False
print(trie.isValidWord("card"))  # True

# Time spent: ~35 minutes